File: python3/environment/pendulum.py
import abc
import config
from math import sin, cos, pi
from typing import List, Tuple


class Pendulum():
    """倒立振子の実験環境です。"""

    def __init__(self, cfg: config.Config):
        super().__init__()

        self._actions = (float(cfg.cfg["ENV_ACTION_LEFT"]),
                         float(cfg.cfg["ENV_ACTION_RIGHT"]))

        self._theta_bounds = (
            float(cfg.cfg["ENV_THETA_LEFT"]), float(cfg.cfg["ENV_THETA_RIGHT"]))
        self._thetadot_bounds = (
            float(cfg.cfg["ENV_THETADOT_LEFT"]), float(cfg.cfg["ENV_THETADOT_RIGHT"]))

        self._theta_space = int(cfg.cfg["ENV_THETA_SPACE"])
        self._thetadot_space = int(cfg.cfg["ENV_THETADOT_SPACE"])

        self._g = float(cfg.cfg["ENV_G"])
        self._l = float(cfg.cfg["ENV_LENGTH"])
        self._m = float(cfg.cfg["ENV_MASS"])

        self._fps = int(cfg.cfg["ENV_FPS"])
        self._tau = 1. / self._fps

        self._init_state = (pi, 0.)
        self._s = self._init_state  # (theta, thetadot)

        self.reset()

    def s_space(self) -> int:
        """s のインデックス取りうる個数を返します。"""
        return self._theta_space * self._thetadot_space

    def a_space(self) -> int:
        """a のインデックスの取りうる個数を返します。"""
        return len(self._actions)

    def s(self) -> int:
        """s のインデックスを返します。"""
        theta_idx = self._digitize(
            self._theta_bounds, self._theta_space, self._s[0])
        thetadot_idx = self._digitize(
            self._thetadot_bounds, self._thetadot_space, self._s[1])

        return theta_idx * self._thetadot_space + thetadot_idx

    def _digitize(self, bounds: List[int], num: int, val: float) -> int:
        """[bounds[0], bounds[1]] を境界として num 個のスペースに区切った場合の
        val の番号を返します。"""
        if val < bounds[0]:
            return 0
        elif val >= bounds[1]:
            return num - 1
        width = (bounds[1] - bounds[0]) / (num - 2)
        return int((val - bounds[0]) // width) + 1

    def info(self) -> str:
        """現在の環境の状態についてわかりやすい文字列を提供します。"""
        return ",".join(map(lambda v: "{:.15f}".format(v), self._s))

    def r(self) -> float:
        """s1 で a1 したとき s2 に移った場合の報酬です。"""
        theta = self._s[0]
        return -abs(theta) + pi / 2.

    def reset(self):
        """環境を初期状態に戻します。"""
        self._s = self._init_state

    def run_step(self, a):
        """a を受け取って内部の状態を遷移させます。"""
        u = self._actions[a]

        self._s = self._solve_runge_kutta(self._s, u, self._tau)

    def _solve_runge_kutta(self, s: Tuple[float, float], u: float, dt: float) -> Tuple[float, float]:
        """s で u を入力したときの dt 時間後の s をルンゲクッタで求めます。"""
        k1 = self._differential(s, u)
        s1 = self._solve_euler(s, k1, dt / 2)
        k2 = self._differential(s1, u)
        s2 = self._solve_euler(s, k2, dt / 2)
        k3 = self._differential(s2, u)
        s3 = self._solve_euler(s, k3, dt)
        k4 = self._differential(s3, u)

        snext = tuple(si + (k1i + 2.*k2i + 2.*k3i + k4i) * dt / 6.
                      for si, k1i, k2i, k3i, k4i in zip(s, k1, k2, k3, k4))

        return self._normalize_s(snext)

    def _differential(self, s: Tuple[float, float], u: float) -> Tuple[float, float]:
        """s で u を加えたときの微分。"""
        theta, thetadot = self._s

        l = self._l
        g = self._g
        m = self._m

        thetaddot = g / l * sin(theta) + u / (m * l**2)
        return (thetadot + thetaddot * self._tau, thetaddot)

    def _solve_euler(self, s: Tuple[float, float], sdot: Tuple[float, float], dt: float) -> Tuple[float, float]:
        """s と sdot より dt 後の s をオイラー法で求める。"""
        return (si + sdoti * dt for si, sdoti in zip(s, sdot))

    def _normalize_s(self, s: Tuple[float, float]) -> Tuple[float, float]:
        return ((s[0] + pi) % (2. * pi) - pi, s[1])

    def is_done(self) -> bool:
        """タスクが終了したかどうかを返します。"""
        return False

    def is_success(self) -> bool:
        """タスクが成功したかどうかを返します。"""
        return False

File: rust/src/environment/pendulum.rs
use crate::config::Config;
use crate::environment::Environment;
use crate::{Action, Reward, State};
use std::error;
use std::f64::consts::PI;

pub struct Pendulum {
    actions: Vec<f64>,

    theta_bounds: (f64, f64),
    thetadot_bounds: (f64, f64),
    theta_space: usize,
    thetadot_space: usize,

    g: f64,
    l: f64,
    m: f64,

    tau: f64,

    init_state: St,
    s: St,

    last_thetaddot: f64,
}

impl Pendulum {
    pub fn new(config: &Config) -> Result<Box<Self>, Box<dyn error::Error>> {
        let actions = vec![
            config.get("ENV_ACTION_LEFT")?.parse()?,
            config.get("ENV_ACTION_RIGHT")?.parse()?,
        ];

        let theta_bounds = (
            config.get("ENV_THETA_LEFT")?.parse()?,
            config.get("ENV_THETA_RIGHT")?.parse()?,
        );
        let thetadot_bounds = (
            config.get("ENV_THETADOT_LEFT")?.parse()?,
            config.get("ENV_THETADOT_RIGHT")?.parse()?,
        );
        let theta_space = config.get("ENV_THETA_SPACE")?.parse()?;
        let thetadot_space = config.get("ENV_THETADOT_SPACE")?.parse()?;

        let g = config.get("ENV_G")?.parse()?;
        let l = config.get("ENV_LENGTH")?.parse()?;
        let m = config.get("ENV_MASS")?.parse()?;

        let fps: u32 = config.get("ENV_FPS")?.parse()?;
        let tau = 1. / fps as f64;

        let init_state = (PI / 3., 0.);
        // let init_state = (PI, 0.);
        let s = init_state;

        let last_thetaddot = 0.;

        Ok(Box::new(Pendulum {
            actions,
            theta_bounds,
            thetadot_bounds,
            theta_space,
            thetadot_space,
            g,
            l,
            m,
            tau,
            init_state,
            s,
            last_thetaddot,
        }))
    }

    fn solve_runge_kutta(&self, s: &St, u: f64, dt: f64) -> St {
        let k1 = self.differential(s, u, dt);
        let s1 = self.solve_euler(s, &k1, dt / 2.);
        let k2 = self.differential(&s1, u, dt / 2.);
        let s2 = self.solve_euler(s, &k2, dt / 2.);
        let k3 = self.differential(&s2, u, dt / 2.);
        let s3 = self.solve_euler(s, &k3, dt);
        let k4 = self.differential(&s3, u, dt);

        let next_val = |s, k1, k2, k3, k4| s + (k1 + 2. * k2 + 2. * k3 + k4) * dt / 6.;
        let theta = next_val(s.0, k1.0, k2.0, k3.0, k4.0);
        let theta = normalize(theta);
        let thetadot = next_val(s.1, k1.1, k2.1, k3.1, k4.1);

        (theta, thetadot)
    }

    fn differential(&self, s: &St, u: f64, dt: f64) -> St {
        let (theta, thetadot) = s;

        let l = self.l;
        let g = self.g;
        let m = self.m;

        let thetadot = -g / l * theta.cos() + u * theta / (m * l * l);
        let thetaddot = g / l * theta.sin() + u / (m * l * l);

        (thetadot, thetaddot)
    }

    fn solve_euler(&self, s: &St, sdot: &St, dt: f64) -> St {
        (s.0 + sdot.0 * dt, s.1 + sdot.1 * dt)
    }
}

impl Environment for Pendulum {
    fn s_space(&self) -> usize {
        self.theta_space * self.thetadot_space
    }

    fn a_space(&self) -> usize {
        self.actions.len()
    }

    fn s(&self) -> State {
        let theta_idx = digitize(&self.theta_bounds, self.theta_space, self.s.0);
        let thetadot_idx = digitize(&self.thetadot_bounds, self.thetadot_space, self.s.1);

        theta_idx * self.thetadot_space + thetadot_idx
    }

    fn info(&self) -> String {
        let theta = self.s.0;
        let thetadot = self.s.1;
        format!("{:.15},{:.15}", theta, thetadot)
    }

    fn r(&self) -> Reward {
        let theta = self.s.0;
        -(theta.abs()) + PI / 2.
    }

    fn reset(&mut self) {
        self.s = self.init_state;
        self.last_thetaddot = 0.
    }

    fn run_step(&mut self, a: Action) {
        let u = self.actions[a];
        self.s = self.solve_runge_kutta(&self.s, u, self.tau);
    }

    fn is_done(&self) -> bool {
        false
    }

    fn is_success(&self) -> bool {
        false
    }
}

fn digitize(bounds: &(f64, f64), num: usize, val: f64) -> usize {
    if val < bounds.0 {
        0
    } else if val >= bounds.1 {
        num - 1
    } else {
        let width = (bounds.1 - bounds.0) / (num - 2) as f64;
        ((val - bounds.0) / width) as usize + 1
    }
}

type St = (f64, f64);

fn normalize(theta: f64) -> f64 {
    (theta + 3. * PI) % (2. * PI) - PI
}
